src/Orchestration/Adapter/KubernetesAPI.php

```diff
@@ -0,0 +1,376 @@
+                'namespace' => $this->namespace,
+                'labels' => array_merge($labels, [
+                    'app.kubernetes.io/component' => $name,
+                    //'app.kubernetes.io/created-by' => $hostname,
+                    'app.kubernetes.io/managed-by' => 'utopia-php-orchestration',
+                ]),
+                'annotations' => [],
+            ],
+            'spec' => [
+                'containers' => [[
+                    'imagePullSecrets' => ($this->regCred ? [[
+                        'name' => 'utopia-orchestration-regcred',
+                    ]] : []),
+                    'name' => $name,
+                    'image' => $image,
+                    'imagePullPolicy' => 'IfNotPresent',
+                    'command' => ($entrypoint ? [[$entrypoint]] : []),
+                    'args' => [$command],
+                    'workingDir' => $workdir,
+                    'env' => $vars,
+                ]],
+            ],
+            'volumes' => $volumes,
+            'restartPolicy' => 'Never',
+            'hostname' => $hostname,
+            'setHostnameAsFQDN' => true,
+            //'subdomain' => 'utopia',
+            'resources' => [
+                'limits' => [
+                    'cpu' => floatval($this->cpus),
+                    'memory' => intval($this->memory) * 1e+6, // Convert into bytes
+                ],
+                'requests' => [
+                    'cpu' => floatval($this->cpus),
+                    'memory' => intval($this->memory) * 1e+6, // Convert into bytes
+                ],
+            ],
+        ];
+
+        $body = array_filter($body, function ($value) {
+            return ! empty($value);
+        });
+
+        $result = $this->call('/api/v1/namespaces/'.$this->namespace.'/pods', 'POST', json_encode($body), [
+            'Content-Type: application/json',
+            'Content-Length: '.\strlen(\json_encode($body)),
+        ]);
+
+        if (! in_array($result['code'], [200, 201, 202])) {
+            throw new Orchestration('Failed to create function environment: '.$result['response'].' Response Code: '.$result['code']);
+        }
+
+        $parsedResponse = json_decode($result['response'], true);
+
+        return $parsedResponse['metadata']['name'];
+    }
+
+    /**
+     * Execute Pod
+     *
+     * @param  string[]  $command
+     * @param  array<string, string>  $vars
+     */
+    public function execute(
+        string $name,
+        array $command,
+        string &$output,
+        array $vars = [],
+        int $timeout = -1
+    ): bool {
+        $body = [
+            'Env' => \array_values($vars),
+            'Cmd' => $command,
+            'AttachStdout' => true,
+            'AttachStderr' => true,
+        ];
+
+        // /api/v1/namespaces/:namespace/pods/:name/exec?command=laboris anim ullamco consequat&stderr=true&stdout=true&container=
+        // Connection upgrade needed to websocket
```

Comment:
@TorstenDittmann  Kubernetes adapter needs a websocket client for this method.

Is it fine if the contributor uses some 3rd party library for that? Any suggestion on which one to pick?

Reply:
https://packagist.org/packages/textalk/websocket

We use that one in our main repo 👍🏻